### ISSUE TITLE
Why are properties of Widget (et alt.) not "wrap"ped?

### ISSUE THREAD

**Issue by sacerdot** [garrigue/lablgtk#34]
Dear @garrigue or whoever knows,
I don't understand why the flag "wrap" is not set on some widgets in .props files, with the result that getter and setter for properties are not generated.

Example: class Widget in GtkBase has no wrap flag and all of its properties are unavailable.

In particular, the setter for the "events" property is not and it is necessary to make "drawing_area"s reactive
to button touches, in order to restore functionalities of no-longer-there widgets.

What is the expected way to restore those props? Write stubs by hand? Use "wrap" everywhere?

Thanks,
C.S.C.

**Comment by sacerdot** [comment]
Follow up: look at one of my commits in the pull request "Cairo support for lablgtk" that solves the issue refactoring the code to enable "wrap" for containers.

My feeling is that all classes should have "wrap" enabled and that, in general, the code should be more uniform and auto-generated.

**Comment by garrigue** [comment]
This is a bootstrapping problem: IIRC the generated files depend on GtkBase, so everything there could not be automated. Since GtkBase was already mature when I introduced propcc, this didn't matter at the time.
I'm fine with your splitting out containers. Not sure yet how to do that with GtkWidget...

**Comment by garrigue** [comment]
We tried to generated code directly, rather than props file, so we only got a very raw binding.
You're probably right that generating props files is going to be much more flexible, and allow a smooth transition.
There is still the problem of generating C level bindings, which we did in the previous attempt.

**Comment by sacerdot** [comment]
I have pushed a commit  where I have made your introspection code compile again with recent compilers and I have added my XSLT code (and an attempt to patch propcc) to automatically generate the .props bindings from the same XML file you generate C code from.

Status:
1. your C-code generation compiles, but fails (I don't know if it ever worked)
2. my XSLT code generates a huge .props file with the declaration of all classes in Gtk-3.0.
    It is already quite useful to fix manually the .props that were handwritten.
    However propcc does not work on it because some identifiers are not valid OCaml ones
     (like the ones expected by propcc). I have committed a tentative propcc able to understand them,
     but that code is still broken and not working.

I have committed the stuff anyway since I am likely not to work on it in the next few week(s) and also because somebody could be inspired and try to push the idea in the future.

**Comment by garrigue** [comment]
Are you talking of the code generation in the gtk3 branch?
It used to produce compilable code. At least hello_world.ml did work (both C and ML automatically generated from the gir).

But I'm not sure it should be a priority at this point: properties are easier and lighter to support, and if you really need a method it's not too hard to wrap it by hand.

**Comment by sacerdot** [comment]
Yes, I am talking abour src/introspection in the lablgtk3 branch.

I agree it's not a priority and I won't be working much more on it. I just made a quick experiment and I wanted to make it public. The generted Gtk-3.0.props file is useful anyway because it saves much time in figuring out properties and method names without looking at Gtk documentation.

**Comment by garrigue** [comment]
Then you're talking about something different.
The fully functional generation code is in the src3 subdirectory gtk3 branch, not lablgtk3.
The code in lablgtk3 is actually the same as master, and is older than in gtk3.
